plutus-core/untyped-plutus-core/src/Language/UntypedPlutusCore/Evaluation/Machine/Cek/ExBudgetMode.hs

```diff
@@ -0,0 +1,140 @@
+{-# LANGUAGE DeriveAnyClass        #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE TypeApplications      #-}
+{-# LANGUAGE UndecidableInstances  #-}
+
+{-# LANGUAGE StrictData            #-}
+
+module Language.UntypedPlutusCore.Evaluation.Machine.Cek.ExBudgetMode
+    ( CekExTally (..)
+    , ExBudgetMode (..)
+    , CountingSt (..)
+    , TallyingSt (..)
+    , RestrictingSt (..)
+    , Hashable
+    , counting
+    , tallying
+    , restricting
+    , restrictingEnormous
+    )
+where
+
+import           PlutusPrelude
+
+import           Language.UntypedPlutusCore.Evaluation.Machine.Cek.Internal
+
+import           Language.PlutusCore.Evaluation.Machine.ExBudget
+import           Language.PlutusCore.Evaluation.Machine.ExMemory
+import           Language.PlutusCore.Evaluation.Machine.Exception
+
+import           Control.Lens                                               (ifoldMap)
+import           Control.Monad.Except
+import           Control.Monad.State.Strict
+import           Data.HashMap.Monoidal                                      as HashMap
+import           Data.Hashable                                              (Hashable)
+import           Data.List                                                  (intersperse)
+import qualified Data.Map.Strict                                            as Map
+import           Data.Semigroup.Generic
+import           Data.Text.Prettyprint.Doc
+import           Text.PrettyBy                                              (IgnorePrettyConfig (..))
+
+-- | A budgeting mode to execute an evaluator in.
+data ExBudgetMode st uni fun = ExBudgetMode
+    { _exBudgetModeSpender :: CekBudgetSpender st uni fun  -- ^ A spending function.
+    , _exBudgetModeInitSt  :: st                           -- ^ An initial state.
+    }
+
+-- | Construct an 'ExBudgetMode' out of a function returning a value of the budgeting state type.
+-- The value then gets added to the current state via @(<>)@.
+monoidalBudgeting
+    :: Monoid st => (ExBudgetCategory fun -> ExBudget -> st) -> ExBudgetMode st uni fun
+monoidalBudgeting toSt = ExBudgetMode spender mempty where
+    spender = CekBudgetSpender $ \key budgetToSpend -> modify' (<> toSt key budgetToSpend)
+
+newtype CountingSt = CountingSt ExBudget
```

Comment:
Worth a short haddock on these: in some sense they're entirely determined by their `Semigroup`/`Monoid` instances, but good to say that.